modules/auxiliary/scanner/http/git_scanner.rb

```diff
@@ -0,0 +1,108 @@
+##
+# This module requires Metasploit: http://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'msf/core'
+
+class Metasploit3 < Msf::Auxiliary
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Auxiliary::WmapScanServer
+  include Msf::Auxiliary::Scanner
+  include Msf::Auxiliary::Report
+
+  def initialize
+    super(
+      'Name'        => 'HTTP Git Scanner',
+      'Description' => 'Detect Git disclosure.',
+      'Author'      => ['Nixawk'],
+      'References'  => [
+        ['URL', 'https://github.com/git/git/blob/master/Documentation/technical/index-format.txt']
+      ],
+      'License'     => MSF_LICENSE
+    )
+
+    register_options(
+      [
+        OptString.new('TARGETURI', [true, 'The test path to .git directory', '/.git/']),
+        OptBool.new('GIT_INDEX', [true, 'Check index file in .git directory', true]),
+        OptBool.new('GIT_CONFIG', [false, 'Check config file in .git directory', true]),
+        OptBool.new('GIT_HEAD', [false, 'Check HEAD file in .git directory', true])
+      ]
+    )
+  end
+
+  def req(filename)
+    send_request_cgi(
+      'uri' => normalize_uri(target_uri, filename)
+    )
+  end
+
+  def git_index_parse(resp)
+    return if resp.blank? || resp.length < 12 # A 12-byte header
+    signature = resp[0, 4]
+    return unless signature == 'DIRC'
+
+    version = resp[4, 4].unpack('N')[0].to_i
+    entries_count = resp[8, 4].unpack('N')[0].to_i
+
+    return unless version && entries_count
+    print_good("#{full_uri} (git repo version #{version}) - #{entries_count} files found")
+
+    report_note(
+      host: rhost,
+      port: rport,
+      proto: 'tcp',
+      type: 'git_disclosure',
+      data: { full_uri: full_uri, version: version, entries_count: entries_count }
```

Comment:
[parse_index_file](https://github.com/loveky/git-tools/blob/master/parse_index_file/parse_index_file.rb) can help us dump the information from ```index``` file as follow:

```
Index version : 2                                                                                                                                                        
Entries number: 3982
Found file: .gitignore
  SHA1    : e355d99cede3f1ee5eea002f80559b82266ede73
  stage   : 0
  ctime   : 2013-08-20 12:02:33.000000000
  mtime   : 2013-05-29 21:18:37.000000000
  size    : 2530
Found file: error_log
  SHA1    : 8bdd4efe13ce386bd74dddde65d370ffdda5d386
  stage   : 0
  ctime   : 2013-08-14 19:07:18.000000000
  mtime   : 2013-08-14 19:07:19.000000000
  size    : 1670
```

Secondly, We can also download target files from the directory **```.git```**. SHA1 is useful, and can be used to generate a url which refers to a target file. How to generate the url ? 

```.gitignore```'s  SHA1 hash can be splited into two, 
   - SHA1[:2]: **e3**
   - SHA1[2:]: **55d99cede3f1ee5eea002f80559b82266ede73**

and we can access ```http(s)://VHOST/TARGETURI/.git/objects/e3/55d99cede3f1ee5eea002f80559b82266ede73``` to extrace the content. 


In the end, my advice is to store usefull informations like this:

| *filename* | *sha1* | *ctime* | *mtime* | *filesize* | *url* |
|:-----------|:-------|:--------|:--------|:-----------|:------|
| .gitignore | e355d99cede3f1ee5eea002f80559b82266ede73 | 2013-08-14 19:07:18.000000000 | 2013-05-29 21:18:37.000000000 | 2530 | http(s)://demo/.git/objects/e3/55d99cede3f1ee5eea002f80559b82266ede73 | 

If so many entries in ```index```, maybe it's better ```store_loot``` the information in a csv format.